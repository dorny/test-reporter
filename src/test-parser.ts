import {TestRunResult} from './test-results'

export interface ParseOptions {
  parseErrors: boolean
  workDir?: string
  trackedFiles: string[]
  directoryMapping?: {from: string; to: string}
}

export interface TestParser {
  parse(path: string, content: string): Promise<TestRunResult>
}
