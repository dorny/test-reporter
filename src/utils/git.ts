export async function listFiles(): Promise<string[]> {
  return []
}
