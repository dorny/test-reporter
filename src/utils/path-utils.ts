export function normalizeDirPath(path: string, addTrailingSlash: boolean): string {
  if (!path) {
    return path
  }

  path = normalizeFilePath(path)
  if (addTrailingSlash && !path.endsWith('/')) {
    path += '/'
  }
  return path
}

export function normalizeFilePath(path: string): string {
  if (!path) {
    return path
  }

  return path.trim().replace(/\\/g, '/')
}

export function getBasePath(path: string, trackedFiles: string[]): string | undefined {
  if (trackedFiles.includes(path)) {
    return ''
  }

  for (const file of trackedFiles) {
    const pathParts = path.split('/')
    const originalLength = pathParts.length
    const fileParts = file.split('/')

    while (pathParts.length && pathParts.slice(-1)[0] === fileParts.slice(-1)[0]) {
      pathParts.pop()
      fileParts.pop()
    }

    // we found some matching path parts
    if (pathParts.length !== originalLength) {
      return pathParts.join('/')
    }
  }

  return undefined
}
