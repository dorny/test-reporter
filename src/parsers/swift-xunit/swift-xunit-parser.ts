import {ParseOptions} from '../../test-parser'
import {JavaJunitParser} from '../java-junit/java-junit-parser'

export class SwiftXunitParser extends JavaJunitParser {
  constructor(readonly options: ParseOptions) {
    super(options)
  }
}
