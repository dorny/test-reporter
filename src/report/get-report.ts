import * as core from '@actions/core'
import {TestExecutionResult, TestRunResult, TestSuiteResult} from '../test-results'
import {Align, formatTime, Icon, link, table} from '../utils/markdown-utils'
import {DEFAULT_LOCALE} from '../utils/node-utils'
import {getFirstNonEmptyLine} from '../utils/parse-utils'
import {slug} from '../utils/slugger'

const MAX_REPORT_LENGTH = 65535
const MAX_ACTIONS_SUMMARY_LENGTH = 131072 // 1048576 soon

export interface ReportOptions {
  listSuites: 'all' | 'failed' | 'none'
  listTests: 'all' | 'failed' | 'none'
  baseUrl: string
  onlySummary: boolean
  useActionsSummary: boolean
  badgeTitle: string
}

const defaultOptions: ReportOptions = {
  listSuites: 'all',
  listTests: 'all',
  baseUrl: '',
  onlySummary: false,
  useActionsSummary: true,
  badgeTitle: 'tests'
}

export function getReport(results: TestRunResult[], options: ReportOptions = defaultOptions): string {
  core.info('Generating check run summary')

  applySort(results)

  const opts = {...options}
  let lines = renderReport(results, opts)
  let report = lines.join('\n')

  if (getByteLength(report) <= getMaxReportLength(options)) {
    return report
  }

  if (opts.listTests === 'all') {
    core.info("Test report summary is too big - setting 'listTests' to 'failed'")
    opts.listTests = 'failed'
    lines = renderReport(results, opts)
    report = lines.join('\n')
    if (getByteLength(report) <= getMaxReportLength(options)) {
      return report
    }
  }

  core.warning(`Test report summary exceeded limit of ${getMaxReportLength(options)} bytes and will be trimmed`)
  return trimReport(lines, options)
}

function getMaxReportLength(options: ReportOptions = defaultOptions): number {
  return options.useActionsSummary ? MAX_ACTIONS_SUMMARY_LENGTH : MAX_REPORT_LENGTH
}

function trimReport(lines: string[], options: ReportOptions): string {
  const closingBlock = '```'
  const errorMsg = `**Report exceeded GitHub limit of ${getMaxReportLength(options)} bytes and has been trimmed**`
  const maxErrorMsgLength = closingBlock.length + errorMsg.length + 2
  const maxReportLength = getMaxReportLength(options) - maxErrorMsgLength

  let reportLength = 0
  let codeBlock = false
  let endLineIndex = 0
  for (endLineIndex = 0; endLineIndex < lines.length; endLineIndex++) {
    const line = lines[endLineIndex]
    const lineLength = getByteLength(line)

    reportLength += lineLength + 1
    if (reportLength > maxReportLength) {
      break
    }

    if (line === '```') {
      codeBlock = !codeBlock
    }
  }

  const reportLines = lines.slice(0, endLineIndex)
  if (codeBlock) {
    reportLines.push('```')
  }
  reportLines.push(errorMsg)
  return reportLines.join('\n')
}

function applySort(results: TestRunResult[]): void {
  results.sort((a, b) => a.path.localeCompare(b.path, DEFAULT_LOCALE))
  for (const res of results) {
    res.suites.sort((a, b) => a.name.localeCompare(b.name, DEFAULT_LOCALE))
  }
}

function getByteLength(text: string): number {
  return Buffer.byteLength(text, 'utf8')
}

function renderReport(results: TestRunResult[], options: ReportOptions): string[] {
  const sections: string[] = []
  const badge = getReportBadge(results, options)
  sections.push(badge)

  const runs = getTestRunsReport(results, options)
  sections.push(...runs)

  return sections
}

function getReportBadge(results: TestRunResult[], options: ReportOptions): string {
  const passed = results.reduce((sum, tr) => sum + tr.passed, 0)
  const skipped = results.reduce((sum, tr) => sum + tr.skipped, 0)
  const failed = results.reduce((sum, tr) => sum + tr.failed, 0)
  return getBadge(passed, failed, skipped, options)
}

function getBadge(passed: number, failed: number, skipped: number, options: ReportOptions): string {
  const text = []
  if (passed > 0) {
    text.push(`${passed} passed`)
  }
  if (failed > 0) {
    text.push(`${failed} failed`)
  }
  if (skipped > 0) {
    text.push(`${skipped} skipped`)
  }
  const message = text.length > 0 ? text.join(', ') : 'none'

  let color = 'success'
  if (failed > 0) {
    color = 'critical'
  } else if (passed === 0 && failed === 0) {
    color = 'yellow'
  }
  const hint = failed > 0 ? 'Tests failed' : 'Tests passed successfully'
  const uri = encodeURIComponent(`${options.badgeTitle}-${message}-${color}`)
  return `![${hint}](https://img.shields.io/badge/${uri})`
}

function getTestRunsReport(testRuns: TestRunResult[], options: ReportOptions): string[] {
  const sections: string[] = []
  const totalFailed = testRuns.reduce((sum, tr) => sum + tr.failed, 0)
  if (totalFailed === 0) {
    sections.push(`<details><summary>Expand for details</summary>`)
    sections.push(` `)
  }

  const shouldShowCoverage = testRuns.some(tr => tr.coverage !== undefined)

  const columnNames = ['Report', 'Passed', 'Failed', 'Skipped', 'Time']
  const columnAligns = [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right]

  if (shouldShowCoverage) {
    columnNames.push('Coverage')
    columnAligns.push(Align.Right)
  }

  if (testRuns.length > 0 || options.onlySummary) {
    const tableData = testRuns
      .filter(tr => tr.passed > 0 || tr.failed > 0 || tr.skipped > 0)
      .map(tr => {
        const time = formatTime(tr.time)
        const name = tr.path
        const passed = tr.passed > 0 ? `${tr.passed} ${Icon.success}` : ''
        const failed = tr.failed > 0 ? `${tr.failed} ${Icon.fail}` : ''
        const skipped = tr.skipped > 0 ? `${tr.skipped} ${Icon.skip}` : ''

        if (shouldShowCoverage) {
          const coverage = tr.coverage !== undefined ? tr.coverage + '%' : ''
          return [name, passed, failed, skipped, time, coverage]
        }

        return [name, passed, failed, skipped, time]
      })

    const resultsTable = table(columnNames, columnAligns, ...tableData)
    sections.push(resultsTable)
  }

  if (options.onlySummary === false) {
    const suitesReports = testRuns.map((tr, i) => getSuitesReport(tr, i, options)).flat()
    sections.push(...suitesReports)
  }

  if (totalFailed === 0) {
    sections.push(`</details>`)
  }
  return sections
}

function getSuitesReport(tr: TestRunResult, runIndex: number, options: ReportOptions): string[] {
  const sections: string[] = []
  const suites = options.listSuites === 'failed' ? tr.failedSuites : tr.suites

  if (options.listSuites !== 'none') {
    const trSlug = makeRunSlug(runIndex)
    const nameLink = `<a id="${trSlug.id}" href="${options.baseUrl + trSlug.link}">${tr.path}</a>`
    const icon = getResultIcon(tr.result)
    sections.push(`## ${icon}\xa0${nameLink}`)

    const time = formatTime(tr.time)
    const headingLine2 =
      tr.tests > 0
        ? `**${tr.tests}** tests were completed in **${time}** with **${tr.passed}** passed, **${tr.failed}** failed and **${tr.skipped}** skipped.`
        : 'No tests found'
    sections.push(headingLine2)

    if (suites.length > 0) {
      const shouldShowCoverage = suites.some(s => s.coverage !== undefined)

      const columnNames = ['Test suite', 'Passed', 'Failed', 'Skipped', 'Time']
      const columnAligns = [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right]

      if (shouldShowCoverage) {
        columnNames.push('Coverage')
        columnAligns.push(Align.Right)
      }

      const suitesTable = table(
        columnNames,
        columnAligns,
        ...suites.map((s, suiteIndex) => {
          const tsTime = formatTime(s.time)
          const tsName = s.name
          const skipLink = options.listTests === 'none' || (options.listTests === 'failed' && s.result !== 'failed')
          const tsAddr = options.baseUrl + makeSuiteSlug(runIndex, suiteIndex).link
          const tsNameLink = skipLink ? tsName : link(tsName, tsAddr)
          const passed = s.passed > 0 ? `${s.passed} ${Icon.success}` : ''
          const failed = s.failed > 0 ? `${s.failed} ${Icon.fail}` : ''
          const skipped = s.skipped > 0 ? `${s.skipped} ${Icon.skip}` : ''

          const coverage = s.coverage !== undefined ? s.coverage + '%' : ''

          if (shouldShowCoverage) {
            return [tsNameLink, passed, failed, skipped, tsTime, coverage]
          }

          return [tsNameLink, passed, failed, skipped, tsTime]
        })
      )
      sections.push(suitesTable)
    }
  }

  if (options.listTests !== 'none') {
    const tests = suites.map((ts, suiteIndex) => getTestsReport(ts, runIndex, suiteIndex, options)).flat()

    if (tests.length > 1) {
      sections.push(...tests)
    }
  }

  return sections
}

function getTestsReport(ts: TestSuiteResult, runIndex: number, suiteIndex: number, options: ReportOptions): string[] {
  if (options.listTests === 'failed' && ts.result !== 'failed') {
    return []
  }
  const groups = ts.groups
  if (groups.length === 0) {
    return []
  }

  const sections: string[] = []

  const tsName = ts.name
  const tsSlug = makeSuiteSlug(runIndex, suiteIndex)
  const tsNameLink = `<a id="${tsSlug.id}" href="${options.baseUrl + tsSlug.link}">${tsName}</a>`
  const icon = getResultIcon(ts.result)
  sections.push(`### ${icon}\xa0${tsNameLink}`)

  sections.push('```')
  for (const grp of groups) {
    if (grp.name) {
      sections.push(grp.name)
    }
    const space = grp.name ? '  ' : ''
    for (const tc of grp.tests) {
      const result = getResultIcon(tc.result)
      sections.push(`${space}${result} ${tc.name}`)
      if (tc.error) {
        const lines = (tc.error.message ?? getFirstNonEmptyLine(tc.error.details)?.trim())
          ?.split(/\r?\n/g)
          .map(l => '\t' + l)
        if (lines) {
          sections.push(...lines)
        }
      }
    }
  }
  sections.push('```')

  return sections
}

function makeRunSlug(runIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}`)
}

function makeSuiteSlug(runIndex: number, suiteIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}s${suiteIndex}`)
}

function getResultIcon(result: TestExecutionResult): string {
  switch (result) {
    case 'success':
      return Icon.success
    case 'skipped':
      return Icon.skip
    case 'failed':
      return Icon.fail
    default:
      return ''
  }
}
