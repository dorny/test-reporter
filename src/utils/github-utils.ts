import {createWriteStream} from 'fs'
import * as core from '@actions/core'
import * as github from '@actions/github'
import {GitHub} from '@actions/github/lib/utils'
import * as stream from 'stream'
import {promisify} from 'util'
import got from 'got'
const asyncStream = promisify(stream.pipeline)

export function getCheckRunContext(): {sha: string; runId: number; branch: string} {
  let branch = github.context.ref
  if (branch.startsWith('refs/heads')) branch = branch.slice(11)
  core.info('Branch: ' + branch)

  if (github.context.eventName === 'workflow_run') {
    core.info('Action was triggered by workflow_run: using SHA and RUN_ID from triggering workflow')
    const event = github.context.payload
    if (!event.workflow_run) {
      throw new Error("Event of type 'workflow_run' is missing 'workflow_run' field")
    }
    return {
      sha: event.workflow_run.head_commit.id,
      runId: event.workflow_run.id,
      branch
    }
  }

  const runId = github.context.runId
  if (github.context.payload.pull_request) {
    core.info(`Action was triggered by ${github.context.eventName}: using SHA from head of source branch`)
    const pr = github.context.payload.pull_request
    return {sha: pr.head.sha, runId, branch}
  }

  return {sha: github.context.sha, runId, branch}
}

export async function downloadArtifact(
  octokit: InstanceType<typeof GitHub>,
  artifactId: number,
  fileName: string,
  token: string
): Promise<void> {
  core.startGroup(`Downloading artifact ${fileName}`)
  try {
    core.info(`Artifact ID: ${artifactId}`)

    const req = octokit.rest.actions.downloadArtifact.endpoint({
      ...github.context.repo,
      artifact_id: artifactId,
      archive_format: 'zip'
    })

    const headers = {
      Authorization: `Bearer ${token}`
    }
    const resp = await got(req.url, {
      headers,
      followRedirect: false
    })

    core.info(`Fetch artifact URL: ${resp.statusCode} ${resp.statusMessage}`)
    if (resp.statusCode !== 302) {
      throw new Error('Fetch artifact URL failed: received unexpected status code')
    }

    const url = resp.headers.location
    if (url === undefined) {
      const receivedHeaders = Object.keys(resp.headers)
      core.info(`Received headers: ${receivedHeaders.join(', ')}`)
      throw new Error('Location header was not found in API response')
    }
    if (typeof url !== 'string') {
      throw new Error(`Location header has unexpected value: ${url}`)
    }

    const downloadStream = got.stream(url, {headers})
    const fileWriterStream = createWriteStream(fileName)

    core.info(`Downloading ${url}`)
    downloadStream.on('downloadProgress', ({transferred}) => {
      core.info(`Progress: ${transferred} B`)
    })
    await asyncStream(downloadStream, fileWriterStream)
  } finally {
    core.endGroup()
  }
}

export async function listFiles(octokit: InstanceType<typeof GitHub>, sha: string): Promise<string[]> {
  core.startGroup('Fetching list of tracked files from GitHub')
  try {
    const commit = await octokit.rest.git.getCommit({
      commit_sha: sha,
      ...github.context.repo
    })
    const files = await listGitTree(octokit, commit.data.tree.sha, '')
    return files
  } finally {
    core.endGroup()
  }
}

async function listGitTree(octokit: InstanceType<typeof GitHub>, sha: string, path: string): Promise<string[]> {
  const pathLog = path ? ` at ${path}` : ''
  core.info(`Fetching tree ${sha}${pathLog}`)
  let truncated = false
  let tree = await octokit.rest.git.getTree({
    recursive: 'true',
    tree_sha: sha,
    ...github.context.repo
  })

  if (tree.data.truncated) {
    truncated = true
    tree = await octokit.rest.git.getTree({
      tree_sha: sha,
      ...github.context.repo
    })
  }

  const result: string[] = []
  for (const tr of tree.data.tree) {
    const file = `${path}${tr.path}`
    if (tr.type === 'blob') {
      result.push(file)
    } else if (tr.type === 'tree' && truncated) {
      const files = await listGitTree(octokit, tr.sha || '', `${file}/`)
      result.push(...files)
    }
  }

  return result
}
