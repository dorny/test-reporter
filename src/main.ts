import * as core from '@actions/core'
import * as github from '@actions/github'
import {GitHub} from '@actions/github/lib/utils'

import {LocalFileProvider} from './input-providers/local-file-provider'
import {FileContent} from './input-providers/input-provider'
import {ParseOptions, TestParser} from './test-parser'
import {TestRunResult, TestRunResultWithUrl} from './test-results'
import {getAnnotations} from './report/get-annotations'
import {getReport} from './report/get-report'

import {DotnetTrxParser} from './parsers/dotnet-trx/dotnet-trx-parser'

import {normalizeDirPath, normalizeFilePath} from './utils/path-utils'
import {getCheckRunContext} from './utils/github-utils'
import {Icon} from './utils/markdown-utils'
import {IncomingWebhook} from '@slack/webhook'
import fs from 'fs'
import bent from 'bent'
import {cwd} from 'process'

async function main(): Promise<void> {
  try {
    const testReporter = new TestReporter()
    await testReporter.run()
  } catch (error) {
    if (error instanceof Error) core.setFailed(error)
    else core.setFailed(JSON.stringify(error))
  }
}

class TestReporter {
  readonly artifact = core.getInput('artifact', {required: false})
  readonly name = core.getInput('name', {required: true})
  readonly path = core.getInput('path', {required: true})
  readonly pathReplaceBackslashes = core.getInput('path-replace-backslashes', {required: false}) === 'true'
  readonly reporter = core.getInput('reporter', {required: true})
  readonly listSuites = core.getInput('list-suites', {required: true}) as 'all' | 'failed'
  readonly listTests = core.getInput('list-tests', {required: true}) as 'all' | 'failed' | 'none'
  readonly maxAnnotations = parseInt(core.getInput('max-annotations', {required: true}))
  readonly failOnError = core.getInput('fail-on-error', {required: true}) === 'true'
  readonly failOnEmpty = core.getInput('fail-on-empty', {required: true}) === 'true'
  readonly workDirInput = core.getInput('working-directory', {required: false})
  readonly onlySummary = core.getInput('only-summary', {required: false}) === 'true'
  readonly token = core.getInput('token', {required: true})
  readonly slackWebhook = core.getInput('slack-url', {required: false})
  readonly githubEvent = core.getInput('github-event', {required: false})
  readonly resultsEndpoint = core.getInput('test-results-endpoint', {required: false})
  readonly resultsEndpointSecret = core.getInput('test-results-endpoint-secret', {required: false})
  readonly octokit: InstanceType<typeof GitHub>
  readonly context = getCheckRunContext()

  constructor() {
    this.octokit = github.getOctokit(this.token)

    if (this.listSuites !== 'all' && this.listSuites !== 'failed') {
      core.setFailed(`Input parameter 'list-suites' has invalid value`)
      return
    }

    if (this.listTests !== 'all' && this.listTests !== 'failed' && this.listTests !== 'none') {
      core.setFailed(`Input parameter 'list-tests' has invalid value`)
      return
    }

    if (isNaN(this.maxAnnotations) || this.maxAnnotations < 0 || this.maxAnnotations > 50) {
      core.setFailed(`Input parameter 'max-annotations' has invalid value`)
      return
    }
  }

  async run(): Promise<void> {
    if (this.workDirInput) {
      core.info(`Changing directory to '${this.workDirInput}'`)
      process.chdir(this.workDirInput)
    }

    core.info(`Check runs will be created with SHA=${this.context.sha}`)

    // Split path pattern by ',' and optionally convert all backslashes to forward slashes
    // fast-glob (micromatch) always interprets backslashes as escape characters instead of directory separators
    const pathsList = this.path.split(',')
    const pattern = this.pathReplaceBackslashes ? pathsList.map(normalizeFilePath) : pathsList

    const inputProvider = new LocalFileProvider(this.name, pattern)

    const parseErrors = this.maxAnnotations > 0
    const trackedFiles = parseErrors ? await inputProvider.listTrackedFiles() : []
    const workDir = this.artifact ? undefined : normalizeDirPath(process.cwd(), true)

    if (parseErrors) core.info(`Found ${trackedFiles.length} files tracked by GitHub`)

    const options: ParseOptions = {
      workDir,
      trackedFiles,
      parseErrors
    }

    core.info(`Using test report parser '${this.reporter}'`)
    const parser = this.getParser(this.reporter, options)

    const results: TestRunResultWithUrl[] = []
    const input = await inputProvider.load()

    if (this.resultsEndpoint?.length > 0) {
      try {
        const readStream = input.trxZip.toBuffer()
        const version = fs.existsSync('test/EVA.TestSuite.Core/bin/Release/version.txt')
          ? fs.readFileSync('test/EVA.TestSuite.Core/bin/Release/version.txt').toString()
          : null
        const commitID = fs.existsSync('test/EVA.TestSuite.Core/bin/Release/commit.txt')
          ? fs.readFileSync('test/EVA.TestSuite.Core/bin/Release/commit.txt').toString()
          : null

        core.info(
          `Using EVA version ${version}, commit ${commitID}, branch ${this.context.branch}, current directory: ${cwd()}`
        )

        const post = bent(this.resultsEndpoint, 'POST', {}, 200)
        await post(
          `TestResults?Secret=${this.resultsEndpointSecret}${version ? '&EVAVersion=' + version : ''}${
            commitID ? '&EVACommitID=' + commitID : ''
          }&EVABranch=${encodeURI(this.context.branch)}`,
          readStream
        )
        core.info(`Uploaded TRX files`)
      } catch (ex) {
        core.warning(`Could not upload TRX ZIP file: ${ex}`)
      }
    }

    for (const [reportName, files] of Object.entries(input.reports)) {
      try {
        core.startGroup(`Creating test report ${reportName}`)
        const tr = await this.createReport(parser, reportName, files)
        if (tr != null) {
          results.push(tr)
        }
      } finally {
        core.endGroup()
      }
    }

    const isFailed = results.some(tr => tr.results.some(r => r.isFailed))
    const conclusion = isFailed ? 'failure' : 'success'
    const passed = results.reduce((sum, tr) => sum + tr.passed, 0)
    const failed = results.reduce((sum, tr) => sum + tr.failed, 0)
    const skipped = results.reduce((sum, tr) => sum + tr.skipped, 0)
    const time = results.reduce((sum, tr) => sum + tr.time, 0)

    core.setOutput('conclusion', conclusion)
    core.setOutput('passed', passed)
    core.setOutput('failed', failed)
    core.setOutput('skipped', skipped)
    core.setOutput('time', time)

    if (results.some(r => r.shouldFail)) {
      core.setFailed(`Failed test were found and the results could not be written to github, so fail this step.`)

      let counter = 0
      core.startGroup('Failed tests')
      for (const r of results.filter(r => r.shouldFail)) {
        for (const rr of r.results) {
          for (const s of rr.failedSuites) {
            for (const g of s.failedGroups) {
              for (const t of g.failedTests) {
                if (++counter > 10) {
                  core.endGroup()
                  return
                }

                core.info(`${t.name}: ${t.error?.message}`)
              }
            }
          }
        }
      }

      core.endGroup()
      return
    }

    if (this.failOnError && isFailed) {
      core.setFailed(`Failed test were found and 'fail-on-error' option is set to ${this.failOnError}`)
      return
    }

    if (results.length === 0 && this.failOnEmpty) {
      core.setFailed(`No test report files were found`)
      return
    }
  }

  async createReport(parser: TestParser, name: string, files: FileContent[]): Promise<TestRunResultWithUrl | null> {
    if (files.length === 0) {
      core.warning(`No file matches path ${this.path}`)
    }

    core.info(`Processing test results for check run ${name}`)

    const results: TestRunResult[] = []
    const result: TestRunResultWithUrl = new TestRunResultWithUrl(results, null)

    for (const {file, content} of files) {
      try {
        core.info(`Processing test results from ${file}`)
        const tr = await parser.parse(file, content)
        results.push(tr)
      } catch (error) {
        core.error(`Processing test results from ${file} failed`)
        throw error
      }
    }

    core.info(`Creating check run ${name}`)
    try {
      const existingChecks = await this.octokit.rest.checks.listForRef({
        ref: this.context.sha,
        ...github.context.repo,
        check_name: name,
        status: 'queued'
      })

      const check =
        existingChecks?.data?.total_count > 0
          ? await this.octokit.rest.checks.get({
              check_run_id: existingChecks.data.check_runs[0].id,
              ...github.context.repo
            })
          : await this.octokit.rest.checks.create({
              head_sha: this.context.sha,
              name,
              status: 'in_progress',
              output: {
                title: name,
                summary: ''
              },
              ...github.context.repo
            })

      if (check.data.status === 'queued') {
        await this.octokit.rest.checks.update({
          check_run_id: check.data.id,
          status: 'in_progress',
          output: {
            title: name,
            summary: ''
          },
          ...github.context.repo
        })
      }

      if (files.length === 0) {
        await this.octokit.rest.checks.update({
          check_run_id: check.data.id,
          status: 'completed',
          output: {
            title: name,
            summary: 'No test result files found'
          },
          conclusion: 'failure',
          ...github.context.repo
        })
        return null
      }

      core.info('Creating report summary')
      const {listSuites, listTests, onlySummary} = this
      const baseUrl = check.data.html_url || ''
      const summary = getReport(results, {listSuites, listTests, baseUrl, onlySummary})

      core.info('Creating annotations')
      const annotations = getAnnotations(results, this.maxAnnotations)

      const isFailed = results.some(tr => tr.result === 'failed')
      const conclusion = isFailed ? 'failure' : 'success'
      const icon = isFailed ? Icon.fail : Icon.success

      core.info(`Updating check run conclusion (${conclusion}) and output`)
      const resp = await this.octokit.rest.checks.update({
        check_run_id: check.data.id,
        conclusion,
        status: 'completed',
        output: {
          title: `${name} ${icon}`,
          summary,
          annotations
        },
        ...github.context.repo
      })
      core.info(`Check run create response: ${resp.status}`)
      core.info(`Check run URL: ${resp.data.url}`)
      core.info(`Check run HTML: ${resp.data.html_url}`)
      core.setOutput('url', resp.data.url)
      core.setOutput('url_html', resp.data.html_url)
      core.info(`Check run details: ${resp.data.details_url}`)
      result.checkUrl = resp.data.html_url

      if (this.slackWebhook && this.context.branch === 'master') {
        const webhook = new IncomingWebhook(this.slackWebhook)
        const passed = results.reduce((sum, tr) => sum + tr.passed, 0)
        const skipped = results.reduce((sum, tr) => sum + tr.skipped, 0)
        const failed = results.reduce((sum, tr) => sum + tr.failed, 0)

        const req = {
          blocks: [
            {
              type: 'section',
              text: {
                type: 'mrkdwn',
                text: `*${name}*`
              }
            },
            {
              type: 'section',
              text: {
                type: 'mrkdwn',
                text: `:large_green_circle: ${passed} :large_orange_circle: ${skipped} :red_circle: ${failed} <${resp.data.html_url}|(view)>`
              }
            }
          ]
        }

        results.map((tr, runIndex) => {
          if (tr.failed === 0) return
          let runName = tr.path.slice(0, tr.path.indexOf('/TestResults/'))
          runName = runName.startsWith('test/') ? runName.slice(5) : runName

          req.blocks.push({
            type: 'section',
            text: {
              type: 'mrkdwn',
              text: `:red_circle: ${tr.failed} in <${resp.data.html_url}#r${runIndex}|${runName}>`
            }
          })

          if (failed <= 10) {
            const failedTests: string[] = []
            tr.failedSuites.map(suite => {
              suite.failedGroups.map(group => {
                group.failedTests.map(test => {
                  failedTests.push(`- <${suite.link}|${test.name}>`)
                })
              })
            })

            req.blocks.push({
              type: 'section',
              text: {
                type: 'mrkdwn',
                text: failedTests.join('\n')
              }
            })
          }
        })

        if (this.githubEvent === 'schedule' || failed > 0) {
          await webhook.send(req)
        }
      }
    } catch (error) {
      core.error(`Could not create check to store the results`)
    }

    return result
  }

  getParser(reporter: string, options: ParseOptions): TestParser {
    return new DotnetTrxParser(options)
  }
}

main()
