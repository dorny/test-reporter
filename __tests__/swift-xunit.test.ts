import * as fs from 'fs'
import * as path from 'path'

import {SwiftXunitParser} from '../src/parsers/swift-xunit/swift-xunit-parser'
import {ParseOptions} from '../src/test-parser'
import {getReport} from '../src/report/get-report'
import {normalizeFilePath} from '../src/utils/path-utils'

describe('swift-xunit tests', () => {
  it('report from swift test results matches snapshot', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'swift-xunit.xml')
    const outputPath = path.join(__dirname, '__outputs__', 'swift-xunit.md')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const trackedFiles = ['Package.swift', 'Sources/AcmeLib/AcmeLib.swift', 'Tests/AcmeLibTests/AcmeLibTests.swift']
    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles
    }

    const parser = new SwiftXunitParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result).toMatchSnapshot()

    const report = getReport([result])
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, report)
  })
})
