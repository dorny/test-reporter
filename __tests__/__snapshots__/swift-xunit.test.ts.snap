// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swift-xunit tests report from swift test results matches snapshot 1`] = `
TestRunResult {
  "path": "fixtures/swift-xunit.xml",
  "suites": [
    TestSuiteResult {
      "groups": [
        TestGroupResult {
          "name": "AcmeLibTests.AcmeLibTests",
          "tests": [
            TestCaseResult {
              "error": undefined,
              "name": "test_always_pass",
              "result": "success",
              "time": 36.386333,
            },
            TestCaseResult {
              "error": undefined,
              "name": "test_always_skip",
              "result": "success",
              "time": 92.039167,
            },
            TestCaseResult {
              "error": {
                "details": undefined,
                "line": undefined,
                "message": undefined,
                "path": undefined,
              },
              "name": "test_always_fail",
              "result": "failed",
              "time": 92.05175,
            },
          ],
        },
      ],
      "name": "TestResults",
      "totalTime": 220.47725000000003,
    },
  ],
  "totalTime": undefined,
}
`;
